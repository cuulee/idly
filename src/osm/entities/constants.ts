export enum Geometry {
  POINT = 'point',
  VERTEX = 'vertex',
  AREA = 'area',
  LINE = 'line',
  RELATION = 'relation'
}
