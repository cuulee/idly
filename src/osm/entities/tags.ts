import { Map } from 'immutable';
export type ITags = Map<string, any>;

export function tagsFactory(obj?: object) {
  return Map(obj);
}
