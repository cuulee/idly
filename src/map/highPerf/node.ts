import { Feature, Point } from 'geojson';
import { Set } from 'immutable';
import { Geometry } from 'osm/entities/constants';
import { Tags } from 'osm/entities/helpers/tags';
import { Node } from 'osm/entities/node';
import { ParentWays } from 'osm/parsers/parsers';
import { presetsMatcher, presetsMatcherCached } from 'osm/presets/presets';
import * as R from 'ramda';
import { weakCache, weakCache2 } from 'utils/weakCache';
type NodeFeat = Feature<Point>;
type NodeGeometry = Geometry.POINT | Geometry.VERTEX | Geometry.VERTEX_SHARED;

interface INodeMarkup {
  icon: string;
  name?: string;
  geometry: NodeGeometry;
}
export const DEFAULT_NODE_ICON = 'circle';

export function nodeCombiner(node: Node, parentWays: ParentWays) {
  return {
    ...nodeToPoint(node),
    properties: {
      ...applyNodeMarkup(
        // presetsMatcher,
        getNodeGeometry(node.id, parentWays),
        node.tags
      ),
      id: node.id
    }
  };
}

export const nodeToPoint = (node: Node): Feature<Point> => {
  return {
    id: node.id,
    type: 'Feature',
    geometry: {
      type: 'Point',
      coordinates: [node.loc.lon, node.loc.lat]
    },
    properties: {}
  };
};

export const applyNodeMarkup = (geometry: NodeGeometry, tags: Tags) => {
  const match = presetsMatcherCached(geometry)(tags);
  return {
    icon: (match && match.icon) || DEFAULT_NODE_ICON,
    name: tags.get('name'),
    geometry
  };
};

/**
 * @NOTE
 *  The way osm works is that it will give you everything inside the bbox
 *   but only a secondary level outside the bbox. So if the a node is
 *   currently a vertex, it could become a vertex_shared in future.
 *   thats why parentWays is updated on every network request.
 */
export function getNodeGeometry(id, parentWays: ParentWays) {
  if (parentWays.has(id))
    return parentWays.get(id).size > 1
      ? Geometry.VERTEX_SHARED
      : Geometry.VERTEX;
  return Geometry.POINT;
}
