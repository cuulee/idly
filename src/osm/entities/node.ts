import { Record } from 'immutable';

import { Geometries } from 'osm/entities/constants';
import { Properties, propertiesGen } from 'osm/entities/helpers/properties';
import { ITags, tagsFactory } from 'osm/entities/helpers/tags';
import { genLngLat, LngLat } from 'osm/geo_utils/lng_lat';

export type Id = string;
export type Version = number;
export type Visible = boolean;

/**
 * @BUG cant use enum Geometries
 *  Need more investigation, for now
 *  converted it to an object
 */
export class Node extends Record({
  id: 'n-0',
  loc: genLngLat([NaN, NaN]),
  tags: tagsFactory(),
  type: 'node',
  properties: propertiesGen()
}) {
  public id: Id;
  public tags: ITags;
  public type: string;
  public loc: LngLat;
  public properties: Properties;
  public set(k: string, v: any): Node {
    return super.set(k, v) as Node;
  }
}

export function nodeFactory(obj: {
  id: Id;
  tags?: ITags;
  loc?: LngLat;
  properties?: Properties;
}) {
  return new Node(obj);
}

console.log(Geometries);
