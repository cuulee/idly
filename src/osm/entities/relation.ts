import { List, Map, Record } from 'immutable';
import { ITags, tagsFactory } from 'osm/entities/helpers/tags';

import { Properties, propertiesGen } from 'osm/entities/helpers/properties';
type Id = string;
type Version = number;
type Visible = boolean;

export class Relation extends Record({
  id: 'w-0',
  properties: propertiesGen(),
  tags: tagsFactory(),
  type: 'relation',
  members: List()
}) {
  public id: Id;
  public type: string;
  public tags: ITags;
  public properties: Properties;
  public members: List<Map<string, any>>;
  public set(k: string, v: any): Relation {
    return super.set(k, v) as Relation;
  }
}

export function relationFactory(obj: {
  id: Id;
  tags?: ITags;
  properties?: Properties;
  members?: List<string>;
}) {
  return new Relation(obj);
}
