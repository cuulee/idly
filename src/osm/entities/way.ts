import { Record, Map, List } from 'immutable';
import { ITags, tagsFactory } from 'osm/others/tags';

import { Properties, propertiesGen } from 'osm/others/properties';

type Id = string;
type Version = number;
type Visible = boolean;
// type Nodes = Map<string, Node>;

var wayBaseRecord = Record({
  id: 'w-0',
  type: 'way',
  tags: tagsFactory(),
  properties: propertiesGen(),
  nodes: List()
});

export class Way extends wayBaseRecord {
  id: Id;
  type: string;
  tags: ITags;
  properties: Properties;
  nodes: List<string>;
  set(k: string, v: any): Way {
    return <Way>super.set(k, v);
  }
}

export function wayFactory(obj: {
  id: Id;
  tags?: ITags;
  properties?: Properties;
  nodes?: List<string>;
}) {
  return new Way(obj);
}
