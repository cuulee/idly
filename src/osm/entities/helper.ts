import { Entity } from 'osm/entities/entities';
import { Node } from 'osm/entities/node';
import { Relation } from 'osm/entities/relation';
import { Way } from 'osm/entities/way';

// export const POINT = 'point';
// export const VERTEX = 'vertex';
// export const AREA = 'area';
// export const LINE = 'line';
// export const RELATION = 'relation';

// type POINT = 'point';
export enum Geometries {
  POINT = 'point',
  VERTEX = 'vertex',
  AREA = 'area',
  LINE = 'line',
  RELATION = 'relation'
}

export function getGeometry(entity: Entity): Geometries {
  if (entity instanceof Node) {
    return isPoi(entity) ? Geometries.POINT : Geometries.VERTEX;
  } else if (entity instanceof Way) {
    return isArea(entity) ? Geometries.AREA : Geometries.LINE;
  } else if (entity instanceof Relation) {
    return isMultipolygon(entity) ? Geometries.AREA : Geometries.RELATION;
  } else {
    throw new Error('unknown type of entity');
  }
}

function isPoi(entity: Node) {
  return true;
}

function isArea(entity: Way) {
  return false;
}

function isMultipolygon(entity: Relation) {
  return false;
}

export function isOnAddressLine(entity: Entity) {
  return false;
  //   return resolver.transient(this, 'isOnAddressLine', function() {
  //     return (
  //       resolver.parentWays(this).filter(function(parent) {
  //         return (
  //           parent.tags.hasOwnProperty('addr:interpolation') &&
  //           parent.geometry(resolver) === 'line'
  //         );
  //       }).length > 0
  //     );
  //   });
}
