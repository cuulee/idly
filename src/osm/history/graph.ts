import { List, Map, Record } from 'immutable';

import { Node } from 'osm/entities/node';
import { Relation } from 'osm/entities/relation';
import { Way } from 'osm/entities/way';
import { Properties, propertiesGen } from 'osm/others/properties';
import { ITags, tagsFactory } from 'osm/others/tags';
import { groupBy } from 'ramda';

export class Graph extends Record({
  node: Map(),
  way: Map(),
  relation: Map()
}) {
  public node: Map<string, Node>;
  public way: Map<string, Way>;
  public relation: Map<string, Relation>;
  public set(k: string, v: any): Graph {
    return super.set(k, v) as Graph;
  }
}
type EntityType = Node | Way | Relation;

export function graphFactory(entitiesList: EntityType[] = []) {
  const createMapFromArray = (array = []) => {
    const map = Map();
    return map.withMutations(m => {
      array.forEach(a => {
        m.set(a.id, a);
      });
    });
  };
  const { node, way, relation } = groupBy(e => e.type, entitiesList);
  return new Graph({
    node: createMapFromArray(node),
    way: createMapFromArray(way),
    relation: createMapFromArray(relation)
  });
}
