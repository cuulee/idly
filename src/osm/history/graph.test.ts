import { List, Map } from 'immutable';
import { wayFactory } from 'src/osm/entities/way';
import { relationFactory } from 'src/osm/entities/relation';
import { nodeFactory, Node } from 'src/osm/entities/node';
import { graphFactory } from 'src/osm/history/graph';
import {
  graphRemoveEntity,
  graphRemoveEntities,
  graphSetEntities,
  graphSetEntity
} from 'src/osm/modifiers/graph.modifiers';

describe('Graph Factory', function() {
  // it('returns a new graph if self is frozen', function () {
  //     var graph = graphFactory();
  //     expect(graph.update()).not.toBe(graph);
  // });
  //   var graph = graphFactory({

  //   });
  var node = nodeFactory({ id: 'n1' });
  var way = wayFactory({ id: 'w-1', tags: Map({ foo: 'bar' }) });
  var relation = relationFactory({ id: 'r1' });
  it('returns self if empty', function() {
    var graph = graphFactory([node, way]);
    expect(graph).toMatchSnapshot();
    expect(graph.node).toEqual(Map({ n1: node }));
    expect(graph.way).toEqual(Map({ 'w-1': way }));
  });

  it('ignore multiple same copies', function() {
    var graph = graphFactory([node, way, node]);
    expect(graph.node.size).toEqual(1);
    expect(graph.way.size).toEqual(1);
  });
  it('ignore multiple same id', function() {
    var n2 = nodeFactory({ id: 'n1', tags: Map({ foo: 'foo' }) });
    var graph = graphFactory([node, way, n2]);
    expect(graph.node.size).toEqual(1);
    expect(graph.way.size).toEqual(1);
  });
  it('adds multiple nodes', function() {
    var n2 = nodeFactory({ id: 'n2', tags: Map({ foo: 'foo' }) });
    var graph = graphFactory([node, way, n2]);
    expect(graph.node.size).toEqual(2);
    expect(graph.way.size).toEqual(1);
  });
  it('adds multiple relations', function() {
    var r1 = relationFactory({ id: 'r1', tags: Map({ foo: 'foo' }) });
    var r2 = relationFactory({ id: 'r2', tags: Map({ foo: 'foo' }) });

    var graph = graphFactory([node, r2, way, r1]);

    expect(graph.node.size).toEqual(1);
    expect(graph.way.size).toEqual(1);
    expect(graph.relation.size).toEqual(2);
  });

  it('adds multiple relations', function() {
    var r1 = relationFactory({ id: 'r1', tags: Map({ foo: 'foo' }) });
    var r2 = relationFactory({ id: 'r2', tags: Map({ foo: 'foo' }) });

    var graph = graphFactory([relation, node, r2, way, r1]);

    expect(graph).toMatchSnapshot();
  });
});

// describe.skip('#remove', function() {
//   const node = nodeFactory({ id: 'n1' });
//   const graph = graphFactory({
//     entities: Map({ n1: nodeFactory({ id: 'n1' }) })
//   });

//   it('returns a new graph', function() {
//     var node2 = nodeFactory({ id: 'n2' });
//     var graph1 = graphSetEntity(graph, node2);
//     expect(graphRemoveEntity(graph1, node2)).toEqual(graph);
//   });

//   it("doesn't modify the receiver", function() {
//     expect(graph.entities.get(node.id)).toEqual(node);
//   });

//   it('removes non existing entity', function() {
//     var node2 = nodeFactory({ id: 'n2' });
//     expect(
//       graphRemoveEntity(graph, node2).entities.get(node2.id)
//     ).toBeUndefined();
//   });

//   it('removes the entity as a parentWay', function() {
//     // var node = new Node({ id: "n" }),
//     var node2 = nodeFactory({ id: 'n2' });
//     var w1 = wayFactory({ id: 'w', nodes: List(['n']) });
//     var graph2 = graphSetEntities(graphFactory({}), List([node2, w1]));

//     expect(graph.remove(w1).parentWays(node)).toEqual([]);
//   });

//   it('removes the entity as a parentRelation', function() {
//     var node = new Node({ id: 'n' }),
//       r1 = Relation({ id: 'w', members: [{ id: 'n' }] }),
//       graph = graphFactory([node, r1]);
//     expect(graph.remove(r1).parentRelations(node)).toEqual([]);
//   });
// });
