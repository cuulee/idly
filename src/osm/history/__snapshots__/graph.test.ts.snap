// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graph Factory adds multiple relations 1`] = `
Object {
  "node": Object {
    "n1": Object {
      "id": "n1",
      "loc": Object {
        "lat": NaN,
        "lon": NaN,
      },
      "properties": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": 0,
        "visible": true,
      },
      "tags": Object {},
      "type": "node",
    },
  },
  "relation": Object {
    "r1": Object {
      "id": "r1",
      "members": Array [],
      "properties": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": 0,
        "visible": true,
      },
      "tags": Object {
        "foo": "foo",
      },
      "type": "relation",
    },
    "r2": Object {
      "id": "r2",
      "members": Array [],
      "properties": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": 0,
        "visible": true,
      },
      "tags": Object {
        "foo": "foo",
      },
      "type": "relation",
    },
  },
  "way": Object {
    "w-1": Object {
      "id": "w-1",
      "nodes": Array [],
      "properties": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": 0,
        "visible": true,
      },
      "tags": Object {
        "foo": "bar",
      },
      "type": "way",
    },
  },
}
`;

exports[`Graph Factory returns self if empty 1`] = `
Object {
  "node": Object {
    "n1": Object {
      "id": "n1",
      "loc": Object {
        "lat": NaN,
        "lon": NaN,
      },
      "properties": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": 0,
        "visible": true,
      },
      "tags": Object {},
      "type": "node",
    },
  },
  "relation": Object {},
  "way": Object {
    "w-1": Object {
      "id": "w-1",
      "nodes": Array [],
      "properties": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": 0,
        "visible": true,
      },
      "tags": Object {
        "foo": "bar",
      },
      "type": "way",
    },
  },
}
`;
